### PR TITLE
Feature/integration test cases

### PR DESCRIPTION
**Pull request by tchataigner**
## Integration test cases

Previously 5 different test cases were discussed after the development of the framework:
- Calling syscalls that may or may not be available.
- Providing malformed or otherwise malicious input to syscalls
- Allocating memory beyond limits.
- Calling specific Wasm instructions.
- Triggering overflows.

The tests added can be found in the `tests` folder of the `integration` module.

### Non available syscalls

WAT based test where we are calling an unknown syscall to the VM. Result are an error pointing us toward that direction.

### Providing malformed or otherwise malicious input to syscalls

Added test is in the form of a Rust crate compiled to wasm at build time, `fil_malformed_syscall`. This crate sends a bad `proof_type` to the `compute_unsealed_sector_cid` syscall.

### Allocating memory beyond limits

Stack overflow was handled on [this commit](https://github.com/filecoin-project/ref-fvm/commit/1183818b1a86190f7df0ed2b3d99f9c7400b6b9b).
### Calling specific wasm instructions

As of now this naming of the test does not really make much sense to me. Please feel free to elaborate on it if you have a better understanding than I do.

### Triggering overflows

Added test in the form of a Rust crate, `fil_integer_overflow` where we create an overflow in an actor's state.

## Follow-up

@raulk @BlocksOnAChain if you want to take a look 👍

### CODE REVIEW

**Comment by mriise** [comment]
integration test actors (`fil_hello_world_actor`, `fil_malformed_syscalls`) should probably end with `_actor` for easy glob selection during building. Generating coverage reports like in #616 and after might use this to exclude/include them when building (as they are built with build.rs instead) eg `cargo build -p "*_actor"`

File: testing/integration/tests/fil-malformed-syscall-actor/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+fvm_sdk = { version = "1.0.0-rc.3", path = "../../../../sdk" }
+fvm_shared = { version = "0.8.0", path = "../../../../shared" }
+
+[build-dependencies]
+wasm-builder = "3.0.1"
+wasmtime = "0.33.0"
```
Comment (Stebalien) [inline testing/integration/tests/fil-malformed-syscall-actor/Cargo.toml:12]: can remove this dep.

File: ipld/encoding/CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,8 @@ Changes to the FVM's shared encoding utilities.
 
 ## [Unreleased]
 
+Changed visibility of `errors` mod based on `testing` feature.
```
[the remark this answers is not in the document]
Comment (tchataigner) [reply ipld/encoding/CHANGELOG.md]: Removed changelog entries and changes in `ipld/encoding/src/lib.rs`

File: sdk/CHANGELOG.md [hunk 3]
```diff
@@ -74,4 +74,4 @@ https://github.com/filecoin-project/fvm-specs/issues/63
 
 ## 0.5.0 [2022-04-11]
 
-Upgrades the SDK to fvm_shared 0.5.0. This release includes a significant breaking change to exit codes.
+Upgrades the SDK to fvm_shared 0.5.0. This release includes a significant breaking change to exit codes.
```
[the remark this answers is not in the document]
Comment (tchataigner) [reply sdk/CHANGELOG.md]: Did a revert, changes applied when I merged master. Weird.

File: testing/integration/tests/fil-integer-overflow-actor/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+fvm_ipld_blockstore = { version = "0.1.1", path = "../../../../ipld/blockstore" }
+
+anyhow = "1.0.57"
+cid = { version = "0.8.2", default-features = false }
+serde = { version = "1.0.136", features = ["derive"] }
+serde_tuple = "0.5"
+
+[build-dependencies]
+wasm-builder = "3.0.1"
+wasmtime = "0.33.0"
```
[the remark this answers is not in the document]
Comment (tchataigner) [reply testing/integration/tests/fil-integer-overflow-actor/Cargo.toml]: Removed it!

File: testing/integration/tests/fil-malformed-syscall-actor/Cargo.toml (hunk 1 above)
Comment (tchataigner) [reply testing/integration/tests/fil-malformed-syscall-actor/Cargo.toml]: Removed it!

File: testing/integration/tests/fil-malformed-syscall-actor/rust-toolchain [hunk 5]
```diff
@@ -0,0 +1 @@
+nightly
```
Comment (mriise) [inline testing/integration/tests/fil-malformed-syscall-actor/rust-toolchain:1]: nit: toolchain versions should be locked at the same version as the rest of the project `nightly-2022-05-09`

File: testing/integration/tests/fil-integer-overflow-actor/rust-toolchain [hunk 6]
```diff
@@ -0,0 +1 @@
+nightly
```
Comment (mriise) [inline testing/integration/tests/fil-integer-overflow-actor/rust-toolchain:1]: same here `nightly-2022-05-09`

**Review by mriise — APPROVED** [review]
LGTM, only nit is to lock the toolchain version to the same as the rest of the project

**Pull request merged**